### PR TITLE
Add support for contenteditable elements in Dropdowns -- Addresses: #19949

### LINKED ISSUE #19949: Dropdown.prototype.keydown prevents focus and swallows some keystrokes in contenteditable elements
The click event will fall through and close dropdowns that contain contenteditables.
The space character will fall through and be cancelled in contenteditable elements in dropdowns.

Issue:
http://codepen.io/anon/pen/oxraRw

PR:
https://github.com/twbs/bootstrap/pull/19950

### PR DESCRIPTION
**Pull request by bo01ean**
Add support for contenteditable elements in Dropdowns
Update tests for contenteditable field editing in Dropdown
Fixes #19949.

---

This change is [Reviewable](https://reviewable.io/reviews/twbs/bootstrap/19950)

### CODE REVIEW

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: a118336241a3e241df926dbfc4a4dc8768ac8bbf
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/131767132

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
As @twbs-savage points out, your tests seem to be breaking in IE8.
See the screencast on https://saucelabs.com/jobs/1cc944cc3c33479cbf07b910e36789ce

File: js/dropdown.js [hunk 1]
```diff
@@ -34,6 +34,13 @@
     return $parent && $parent.length ? $parent : $this.parent()
   }
 
+  // INPUT TYPE CHECK
+  // ================
+  function IsEditableElement(e) {
+    return /input|textarea/i.test(e.target.tagName)
+      || e.target.attributes.getNamedItem('contenteditable')
```
Comment (cvrebert) [inline js/dropdown.js]: Why `.attributes.getNamedItem` rather than `.hasAttribute` ?

File: js/tests/unit/dropdown.js [hunk 2]
```diff
@@ -342,19 +343,32 @@ $(function () {
...
+    var keys = [38,40,27,32]
 
     $dropdown
       .parent('.dropdown')
       .on('shown.bs.dropdown', function () {
         assert.ok(true, 'shown was fired')
 
-        $input.trigger('focus').trigger($.Event('keydown', { which: 38 }))
-        assert.ok($(document.activeElement).is($input), 'input still focused')
+        keys.forEach(function (key) {
```
Comment (cvrebert) [inline js/tests/unit/dropdown.js]: Don't use `Array.forEach`. It's not supported in all the browsers that Bootstrap 3 supports.

File: js/dropdown.js (hunk 1 above)
Comment (cvrebert) [inline js/dropdown.js]: Actually, what about https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/isContentEditable ? That seems to handle descendants of contentEditable elements properly.

**Comment by cvrebert** [comment]
Also, excluding the `dist/` and `assets/` files from your commits would be appreciated. They lead to unnecessary merge conflicts.

File: js/tests/unit/dropdown.js (hunk 2 above)
Comment (bo01ean) [inline js/tests/unit/dropdown.js]: This is why @twbs-savage failed above.

File: js/dropdown.js (hunk 1 above)
Comment (bo01ean) [inline js/dropdown.js]: Good idea

**Comment by bo01ean** [comment]
I closed and re-opened for rebuild, trying:
@twbs-savage retry

**Comment by cvrebert** [comment]
@twbs-savage retry

**Comment by bo01ean** [comment]
@twbs-savage retry

**Comment by mdo** [comment]
[**Bootstrap 3 is no longer being officially developed or supported.**](https://github.com/twbs/bootstrap/issues/20631)

All work has moved onto our next major release, v4. As such, this issue or pull request is being closed as a "won't fix." For additional help and support, we recommend [utilizing our community resources](https://github.com/twbs/bootstrap/blob/master/README.md#community). Thanks for your understanding, and see you on the other side of v4!

<3,
@mdo and team

**Pull request closed without merging**